docs/developers/arch/adr-003.md

```diff
@@ -0,0 +1,114 @@
+
+This document describes a proposed plan to clarify the structure of the Hypothesis client's sidebar application.
+
+## Context
+
+The Hypothesis client consists of three major pieces:
+
+- The "boot" code which loads resources required by other parts of the client
+- The "annotator" code which runs in the context of the host page
+- The "sidebar" application which handles fetching, presenting and editing of
+  annotations
+
+The sidebar application is built on Angular JS [1]. Other than the [components
+ADR](adr-001.md) the architecture has evolved in a slightly ad-hoc fashion.
+
+For example, historically the app suffered from a number of problems related to
+performance [2] and testability [3] which motivated the introduction of a
+central state container. However it is not universally used for all the data
+needed to display the UI, creating confusion about what to use for new features.
+
+In addition to inconsistencies, the structure that exists is difficult to
+discern without extensive reading of the code and new or occassional developers
+have found it difficult to grok. Poor naming of some modules also complicates
+understanding. For example the API client is called `store.js` (a reference to
+_Annotator Store_).
+
+## Proposal
+
+Make explicit the implicit structure which the sidebar app's code has gradually
+been evolving toward over the last 12-18 months and to finish the transition.
+The sidebar app will consist of the following layers:
+
+**Store**
+- The _state container_ [4] for the app, built on Redux.
+- This maintains all the state needed to "draw" the user interface in a normalized (ie. avoiding duplication) representation.
+- Organised into modules which maintain different aspects of the state.
+
+**Services**
+- Classes that handle interaction with external actors such as the host page, the annotation service (via the API, real-time API and OAuth endpoints) and local storage, scheduled/periodic tasks.
+- These may maintain transient state (eg. the active `WebSocket` instance), but nothing required to "draw" the UI
+
+**Components**
+- Self-contained parts of the UI, each consisting of a template, styles and logic.
+- These are connected to the services and store.
+
+**Utilities**
+- Helper functions and classes which do not depend on the other layers.
+
+These layers will have the following dependency structure:
+
+```
+Utilities --> Store
+   |          | | |
+   |          V | |
+   |-------> Services
+   |          | | | |
+   |          V V | |
+   |-------> Components
+   |          |   | |
+   |          V   V V
+   |-------> [ Entry point ]
+```
+
+The directory structure will be organized along functional lines in a similar
+manner to our Python apps:
+
+```
+src/sidebar/
+  store/ # Redux store
+    modules/ # Modules defining the logic related to a given part of the state
+             # (eg. "annotations", "groups", "drafts").
+  services/
+  components/
+  util/ # Utility classes & functions
```

Comment:
These directories aren't in alphabetical order here, presumably because you are trying to put more attention on certain directories, but that could be maybe a little confusing?

Reply:
Ah noted. I was indeed putting more emphasis on the "store" part because that's where clarity is missing most at the moment.